### PR TITLE
Encoder improvements and  Fix indexPointsMultiplier

### PR DESCRIPTION
**Pull request by IhorNehrutsa**
59kHz  SystemClock_Config_HSE_8M_SYSCLK_72M()
102kHz SystemClock_Config_HSE_8M_SYSCLK_128M()

### CODE REVIEW

**Comment by CAP1Sup** [comment]
Microstep multiplier was never made to be set like that. It was meant so that the user could set it to a specific value. That way, they could have motors with differing distances using the same step pin.

**Comment by CAP1Sup** [comment]
Why not just use the increments data types for all functions that get increments

**Comment by IhorNehrutsa** [comment]
What mean " increments data types"? Integers?

**Comment by CAP1Sup** [comment]
Sorry, my autocorrect added the s. I meant why not use the `increments` data type for all of the functions that deal with increments?

**Comment by IhorNehrutsa** [comment]
> Why not just use the increments data type for all functions that get increments?

Let's view at this in the future

**Comment by IhorNehrutsa** [comment]
Ready, bugs are possibly. ;)

File: src/hardware/encoder.cpp [hunk 1]
```diff
@@ -824,10 +838,24 @@ void Encoder::setStepOffset(double offset) {
...
+    startupIncrementsOffset = offset;
+}
+
+
 // Set encoder zero point
 void Encoder::zero() {
 
+    // Populate the average angle reading table
+    for (uint8_t index = 0; index < (ANGLE_AVG_READINGS * 2); index++) {
+        getAngleAvg();
```
Comment (CAP1Sup) [inline src/hardware/encoder.cpp:852]: No need to get the angle avg, the averaging is done by the increments.

File: src/hardware/encoder.h [hunk 2]
```diff
@@ -330,17 +343,19 @@ class Encoder {
         int32_t revolutions = 0;
 
         // Moving average instances
-        MovingAverage <float> speedAvg;
-        MovingAverage <int16_t> rawSpeedAvg;
-        MovingAverage <float> accelAvg;
-        MovingAverage <uint16_t> incrementAvg;
-        MovingAverage <float> absAngleAvg;
-        MovingAverage <int16_t> rawTempAvg;
+        MovingAverage <float, float> speedAvg;
```
Comment (CAP1Sup) [inline src/hardware/encoder.h:346]: This was an excellent idea!

File: src/hardware/motor.cpp [hunk 3]
```diff
@@ -92,6 +92,7 @@ float StepperMotor::getEstimRPM() {
 float StepperMotor::getDegreesPS() {
     calc:
     while (isStepping)
+        ; // wait end of stepping calculation
```
Comment (CAP1Sup) [inline src/hardware/motor.cpp:95]: Why put the comma on this line? Can't it be put on the same line as the while? Just minor preference for clean files

File: src/hardware/motor.cpp [hunk 4]
```diff
@@ -126,7 +127,8 @@ int32_t StepperMotor::getStepPhase() {
 
 // Returns the desired angle of the motor
 float StepperMotor::getDesiredAngle() {
-    return (this -> desiredAngle);
+    //return (this -> desiredAngle);
```
Comment (CAP1Sup) [inline src/hardware/motor.cpp:130]: This was a good idea, I was thinking about implementing a similar thing

File: src/hardware/motor.cpp [hunk 5]
```diff
@@ -292,6 +294,28 @@ void StepperMotor::setMicrostepping(uint16_t setMicrostepping) {
 
         // Fix the microsteps per rotation
         this -> microstepsPerRotation = round(360.0 / microstepAngle);
+
+        // Set the microstepMultiplier according to the microstepDivisor
```
Comment (CAP1Sup) [inline src/hardware/motor.cpp:298]: This isn't how the microstep multiplier is supposed to work. I'll leave it in, but I'm going to add it as a config option.

File: src/user/main.cpp [hunk 6]
```diff
@@ -60,6 +60,7 @@ void setup() {
     #ifdef CHECK_ENCODER_SPEED
         while(true) {
             GPIO_WRITE(LED_PIN, HIGH);
+            // #define SYSCLK_FREQ 128 <--
```
Comment (CAP1Sup) [inline src/user/main.cpp:63]: Why have this here?

**Comment by CAP1Sup** [comment]
Almost ready to merge, just need to resolve questions

**Comment by IhorNehrutsa** [comment]
I have two days  sea holiday. See on monday

**Comment by CAP1Sup** [comment]
Enjoy your vacation!

File: src/hardware/motor.cpp (hunk 3 above)
Comment (IhorNehrutsa) [reply src/hardware/motor.cpp:95]: Let use auto cede formatting.   #47

**Comment by IhorNehrutsa** [comment]
I don't understand the abstraction of the microstepMultiplier. 
Which it means?
How to use it?
Is there any example?
See Index points multiplier in fastSine.cpp

**Comment by IhorNehrutsa** [comment]
@CAP1Sup Could we continue to work?

**Comment by CAP1Sup** [comment]
> @CAP1Sup Could we continue to work?
> 
> 

Sorry, I've been busy recently. I'll work on it over the weekend.

**Comment by CAP1Sup** [comment]
All done, sorry for the delay @IhorNehrutsa

**Pull request merged**